youtube_dlc/extractor/niconico.py

```diff
@@ -284,80 +329,79 @@ def _real_extract(self, url):
-                    items = [items]
-                for item in items:
-                    ret = xpath_text(video_info_xml, './/' + item)
-                    if ret:
-                        return ret
-
-            video_real_url = flv_info['url'][0]
-
-            extension = get_video_info('movie_type')
-            if not extension:
-                extension = determine_ext(video_real_url)
-
-            formats = [{
+        # Get video info
+        video_info_xml = self._download_xml(
+            'http://ext.nicovideo.jp/api/getthumbinfo/' + video_id,
+            video_id, note='Downloading video info page')
+
+        def get_video_info_xml(items):
+            if not isinstance(items, list):
+                items = [items]
+            for item in items:
+                ret = xpath_text(video_info_xml, './/' + item)
+                if ret:
+                    return ret
+
+        if get_video_info_xml('error'):
+            error_code = get_video_info_xml('code')
+
+            if error_code == 'DELETED':
+                raise ExtractorError('The video has been deleted.',
+                                     expected=True)
+            elif error_code == 'NOT_FOUND':
+                raise ExtractorError('The video is not found.',
+                                     expected=True)
+            elif error_code == 'COMMUNITY':
+                raise ExtractorError('The video is community members only.',
+                                     expected=True)
+            else:
+                raise ExtractorError('%s reports error: %s' % (self.IE_NAME, error_code))
+
+        # Start extracting video formats
+        formats = []
+
+        # Get HTML5 videos info
+        dmc_info = api_data['video'].get('dmcInfo')
+
+        quality_info = dmc_info['quality']
+        for audio_quality in quality_info['audios']:
+            for video_quality in quality_info['videos']:
+                if not audio_quality['available'] or not video_quality['available']:
+                    continue
+                formats.append(self._extract_format_for_quality(
+                    api_data, video_id, audio_quality, video_quality))
+
+        # Get flv/swf info
+        video_real_url = api_data['video']['smileInfo']['url']
+        is_economy = video_real_url.endswith('low')
+
+        extension = get_video_info_xml('movie_type')
+        if not extension:
+            extension = determine_ext(video_real_url)
+
+        # Economy mode movie is not source movie.
+        # Similarly, if movie file size is unstable, old server movie is not source movie.
+        if not is_economy and int(get_video_info_xml('size_high')) > 1:
+            formats.append({
                 'url': video_real_url,
+                'format_id': 'smile',
+                'format_note': 'Source movie file in old server (SMILEVIDEO)',
+                'preference': -2,  # I think demand for source movie is not high.
```

Comment:
Please explain. This should probably be `source_preference`

Reply:
**Summary: Most of movie files in SMILEVIDEO (old server) before 2018-06-28 are SOURCE files. But, not recommend for obsolete encoding and lack of movie file information.**

According to official and unofficial technical documents written by Japanese.

Until 2018-06-28, under the following conditions SMILEVIDEO preserved uploaded movies without change.

- Movie encoding is swf, flv (H.263 / VP6), or mp4 (H.264).
- Movie size is below 100MB (uploaded by paid member) or 40MB (free member).
- Etc. Etc. (Japanese information : http://nicowiki.com/?cmd=backup&page=%E3%82%A8%E3%83%B3%E3%82%B3%E3%83%BC%E3%83%89%E8%A8%AD%E5%AE%9A&age=86#fe0ad87d )

The other cases, uploaded movies were re-encoded to mp4 (H.264).

However, SMILEVIDEO is deprecated system. So, download SMILEVIDEO movie file have several problems.

- Movie encoding : Now, swf and flv are obsolete. DMC (new server) always re-encoded them to mp4 (H.264).
- Reducing API : Now, information of SMILEVIDEO movie file from API are URL, extension and filesize only. API for DMC provided resolution and bitrate, mandatory information to detect download movie file.

Because I down 'preference', not 'source_preference'.

Reply:
ok, makes sense

Reply:
@kurumigi this is wrong. `smile_high` source files are the most important download from niconico because for the time period you specify (which might actually be around november 2016, i don't remember what my notes said), the smile_high videos are the highest quality for both video *and* audio.

The DMC streams are always worse in this regard, with degraded audio and reencoded video (which is usually downscaled to the nearest 16:9 standard resolution). The fact that some files use outdated codecs doesn't matter; any modern video player (VLC, MPC, MPV) can play them without issue. Since youtube-dl is an archival tool, the highest quality version should be preferred regardless. If you need certain codecs you can reencode them yourself, and you will still end up with a better quality video than the DMC versions because niconico uses very low quality settings for conversion.

There is no reason why they shouldn't be preferred (for videos uploaded before 2018 or 2016, need to double check). There are plenty of videos with 320kbit audio being reencoded to 192kbit in the DMC streams which is horrible. I can give plenty of examples of videos where quality is degraded from DMC conversion

Reply:
@kurumigi
`quality` is used to mark that a video is better than ALL videos of a higher `quality`
`source_quality` is used to indicate that the format has a source that is prefered over other sources (this does not significantly affect the actual format selection unless explicitly specified by the user with `-S source`)

Like I said before, I don't know anything about niconico. But if @bbepis is correct, the quality should not be set lower and instead `source_quality` should be set lower (since less metadata is available in the older source and therefore, the newer source is preferred if everything else is equal)

All this is assuming that the video codec and resolution (or bitrate) of the formats is known. If these cannot be determined, the situation is more complicated.

Please give me an example output of `-list-formats` of a video with both sources so that I can better understand the situation

Reply:
@pukkandan The main thing to grasp is that there are two distinct sources available for videos; `DMC` and `smile`.

Originally there was only the `smile` format; it would come in two variants, either `smile_high` or `smile_low` depending on if the website is currently under economy hours. This was the source video that was uploaded to NND, because niconico would cap uploads at 100MB

However sometime during 2016 (or 2018 according to that wiki linked), they introduced a new video serving system named `DMC` that would serve videos with HLS and standardized bitrates / resolutions. During this time they also removed the 100MB cap because everything would get reencoded regardless.

This is why the date is important; for videos uploaded before DMC was introduced, `smile_high` was the highest quality because it was the *original* video that was uploaded, and DMC sources are retroactively encoded from that. After that date, `smile_high` isn't necessarily the highest quality anymore because it's just a reencode designed to mimic the old limits, so the regular format selection rules apply to those.

I wrote something [more indepth in a reddit comment](https://www.reddit.com/r/youtubedl/comments/kluzux/i_have_a_fork_of_youtubedl_with_a_lot_of_fixes/gju6wob/?utm_source=reddit&utm_medium=web2x&context=3) but it focuses more on the history of the site

[Here is my logic](https://github.com/animelover1984/youtube-dl/blob/master/youtube_dl/extractor/niconico.py#L436-L491) that I use for `smile` format extraction, feel free to copy it outright.

I make use of ffprobe on the `smile` video to determine pretty much all metadata about the video because as of december 2020 the niconico API no longer contains any video info on the smile format for some reason.

Here is an example of the format selection that my fork gives for https://www.nicovideo.jp/watch/sm23133906:


![image](https://user-images.githubusercontent.com/36346617/106915991-cb45d100-675a-11eb-8c0d-137fdf3118c9.png)

Reply:
Here are some format selection results for SWF and FLV formats:
![image](https://user-images.githubusercontent.com/36346617/106917270-36dc6e00-675c-11eb-841d-d6bcc83b29da.png)

![image](https://user-images.githubusercontent.com/36346617/106917470-668b7600-675c-11eb-9880-7f3913814517.png)

The API does not even tell you the format of the video, so ffprobe is required to even determine if it's an mp4, flv or swf file.

As I've said before if the formats are not suitable for the end user, they can either set `-f` to only grab the mp4/DMC variants, or use the remux post-processor to remux it to mkv (where it can then be played in any decent video player)

Reply:
Is there always only 1 smile format? Will it always be in the highest resolution?

Reply:
> Is there always only 1 smile format? Will it always be in the highest resolution?

`smile_high` is the higher quality, compared to `smile_low`. `smile_high` will only show if the user has a premium account, or it will show for free users of the site during offpeak hours (around 12 AM JST - 10 AM JST, they might have changed it).

From what I understand it's possible to show both formats in youtube-dl (assuming the user has access to the higher quality version), however I haven't figured out how to do it. Mainly because there wouldn't be any practical benefit, as the default version is the highest quality smile version that is available.

Reply:
Here's an example of a video uploaded in 2020 (after DMC was introduced) where the highest quality DMC video is better than `smile_high`:

![image](https://user-images.githubusercontent.com/36346617/106922734-72c60200-6761-11eb-8849-5b131e5afca8.png)

Reply:
Hmm... Using ffprobe is an interesting idea.

Apart fromt hat,  API return "filesize is 1 byte" like this, the movie no longer stored in SMILEVIDEO.
This matter happens only in movies uploaded after 2018-06-28, it's one of the reasons I think "deadline" is 2018-06-28.

Reply:
now only this issue left

Reply:
I re-check official and unofficial technical documents written by Japanese, I understand **major** specification changes of niconico douga are two times.

- **Movies uploaded before 2016-12-08**:
    **Maximum Movie filesize was 100MB**.
    SMILEVIDEO preserved acceptable encode (wrote in previous comment) movies, and re-encoded unacceptable movies to H.264.
    After 2018-04-10, DMC re-encoded all movies from SMILEVIDEO. (Official Source: http://blog.nicovideo.jp/niconews/69409.html (Japanese))
    
    **Conclusion : SMILEVIDEO's movie are always superior to DMC's one.**

- **Movies uploaded between 2016-12-08 and 2018-06-28**:
    **Maximum Movie filesize was upped to 1.5GB.**
    SMILEVIDEO preserved acceptable encode movies **below 100MB**, and re-encoded unacceptable movies to H.264 **below 100MB**. (Archive of Official Source: https://www.nicovideo.jp/watch/sm33168111 (Japanese / Movie))
    DMC re-encoded all movies immediately.
    
    **Conclusion : SMILEVIDEO acceptable movies are superior to DMC's one. Other movies depends re-encoded movie size.**
    
- **Movies uploaded after 2018-06-28**:
    SMILEVIDEO deactivated. No longer stored movies. (Official Source: https://twitter.com/nico_nico_talk/status/1004290141530173440 (Japanese))
    DMC re-encoded all movies immediately.
    
    **Conclusion : DMC's movie only.**

Hmmmm....

Reply:
@bbepis

> Here is my logic that I use for smile format extraction, feel free to copy it outright.

 I will merge this code. Thank you for permission to copy code.

Reply:
@bbpeis Could you review the metadata extraction. I don't fully understand it